core/src/main/scala/org/apache/spark/CacheRecoveryManager.scala

```diff
@@ -0,0 +1,189 @@
+ * it down.
+ */
+final private class CacheRecoveryManager(
+    blockManagerMasterEndpoint: RpcEndpointRef,
+    executorAllocationManager: ExecutorAllocationManager,
+    conf: SparkConf)
+  extends Logging {
+
+  private val forceKillAfterS = conf.get(DYN_ALLOCATION_CACHE_RECOVERY_TIMEOUT)
+  private val threadPool = ThreadUtils.newDaemonCachedThreadPool("cache-recovery-manager-pool")
+  private implicit val asyncExecutionContext: ExecutionContext =
+    ExecutionContext.fromExecutorService(threadPool)
+  private val scheduler =
+    ThreadUtils.newDaemonSingleThreadScheduledExecutor("cache-recovery-shutdown-timers")
+  private val recoveringExecutors: mutable.Set[String] =
+    ConcurrentHashMap.newKeySet[String]().asScala
+
+  /**
+   * Start the recover cache shutdown process for these executors
+   *
+   * @param execIds the executors to start shutting down
+   */
+  def startCacheRecovery(execIds: Seq[String]): Unit = {
+    logDebug(s"Recover cached data before shutting down executors ${execIds.mkString(", ")}.")
+    val canBeRecovered = checkMem(execIds)
+    recoveringExecutors ++= canBeRecovered
+    val executorsWithKillTimers = canBeRecovered.zip(canBeRecovered.map(startKillTimer))
+    executorsWithKillTimers.foreach((replicateUntilDone _).tupled)
+  }
+
+  /**
+   * Given a list of executors that will be shut down, check if there is enough free memory on the
+   * rest of the cluster to hold their data. Return a list of just the executors for which there
+   * will be enough space. Executors are included smallest first.
+   *
+   * @param execIds executors which will be shut down
+   * @return a Seq of the executors we do have room for
+   */
+  private def checkMem(execIds: Seq[String]): Seq[String] = {
+    val execsToShutDown = execIds.toSet
+    // Memory Status is a map of executor Id to a tuple of Max Memory and remaining memory on that
+    // executor.
+    val allExecMemStatus: Map[String, (Long, Long)] = blockManagerMasterEndpoint
+      .askSync[Map[BlockManagerId, (Long, Long)]](GetMemoryStatus)
+      .map { case (blockManagerId, mem) => blockManagerId.executorId -> mem }
+
+    val (expiringMemStatus, remainingMemStatus) = allExecMemStatus.partition {
+      case (execId, _) => execsToShutDown.contains(execId)
+    }
+    val freeMemOnRemaining = remainingMemStatus.values.map(_._2).sum
+
+    // The used mem on each executor sorted from least used mem to greatest
+    val executorAndUsedMem: Seq[(String, Long)] =
+      expiringMemStatus.map { case (execId, (maxMem, remainingMem)) =>
+        val usedMem = maxMem - remainingMem
+        execId -> usedMem
+      }.toSeq.sortBy { case (_, usedMem) => usedMem }
+
+    executorAndUsedMem
+      .scan(("start", freeMemOnRemaining)) {
+        case ((_, freeMem), (execId, usedMem)) => (execId, freeMem - usedMem)
+      }
+      .drop(1)
+      .filter { case (_, freeMem) => freeMem > 0 }
```

Comment:
added comment

Reply:
conversely, you might decide to not even try to replicate some executors, because *all* of their data can't be replicated, but in fact *some* of the blocks on the executor could be replicated (maybe even the most useful ones).

given the limitations & complications of this method, I'm wondering whether its even worth it to do this filtering?

Reply:
That's a good point. I was really concerned with not hurting the stability of the cluster. So in the case where we are running out of cluster memory, I thought it would be safer to fail fast and give up on replication. 

Your right it would probably be more useful to get as many complete RDDs recovered as possible, and this method is complicated. I can test taking out this check and see what happens.